Implement a few gas optimizations

scorpion9979 opened the pull request:
Addresses both https://github.com/Kwenta/token/issues/244 and https://github.com/Kwenta/token/issues/245. Below is a gas snapshot comparison highlighting "claimEpoch" and "claimMany" functions:

Before:
> $ forge snapshot --match-contract TokenDistributor --match-test testClaim
[⠢] Compiling...No files changed, compilation skipped
[⠆] Compiling...
Running 14 tests for test/foundry/unit/TokenDistributor/TokenDistributor.t.sol:TokenDistributorTest
[PASS] testClaimAlreadyClaimed() (gas: 362543)
[PASS] testClaimEpoch() (gas: 366524)
[PASS] testClaimEpoch0() (gas: 996862)
[PASS] testClaimEpoch3People() (gas: 1317344)
[PASS] testClaimEpochAhead() (gas: 14524)
[PASS] testClaimEpochCheckpoint() (gas: 537537)
[PASS] testClaimEpochNotReady() (gas: 30835)
[PASS] testClaimEpochVestingEntry() (gas: 360584)
[PASS] testClaimMany() (gas: 447099)
[PASS] testClaimMultipleClaims() (gas: 657686)
[PASS] testClaimNoEpochYet() (gas: 14491)
[PASS] testClaimNoStaking() (gas: 166608)
[PASS] testClaimNotStaker() (gas: 393185)
[PASS] testClaimPreviouslyStaked() (gas: 456505)
Test result: ok. 14 passed; 0 failed; 0 skipped; finished in 7.36ms

After:
> $ forge snapshot --match-contract TokenDistributor --match-test testClaim
[⠆] Compiling...
[⠊] Compiling 7 files with 0.8.19
[⠢] Solc 0.8.19 finished in 9.54sCompiler run successful!
[⠆] Solc 0.8.19 finished in 9.54s
Running 14 tests for test/foundry/unit/TokenDistributor/TokenDistributor.t.sol:TokenDistributorTest
[PASS] testClaimAlreadyClaimed() (gas: 362663)
[PASS] testClaimEpoch() (gas: 366626)
[PASS] testClaimEpoch0() (gas: 1052496)
[PASS] testClaimEpoch3People() (gas: 1373200)
[PASS] testClaimEpochAhead() (gas: 14530)
[PASS] testClaimEpochCheckpoint() (gas: 537747)
[PASS] testClaimEpochNotReady() (gas: 30838)
[PASS] testClaimEpochVestingEntry() (gas: 360686)
[PASS] testClaimMany() (gas: 421716)
[PASS] testClaimMultipleClaims() (gas: 614312)
[PASS] testClaimNoEpochYet() (gas: 14497)
[PASS] testClaimNoStaking() (gas: 166710)
[PASS] testClaimNotStaker() (gas: 393392)
[PASS] testClaimPreviouslyStaked() (gas: 456601)
Test result: ok. 14 passed; 0 failed; 0 skipped; finished in 5.37ms

The changes actually increase the gas cost a bit for single claims via `claimEpoch`, but render batch claims via `claimMany` more efficient.

scorpion9979 commented:
@jcmonte I noticed that commit https://github.com/Kwenta/token/pull/246/commits/f8c9520a1e829ea9bcefd402bf4e60a67ea35569 now breaks [TokenDistributor.t.sol](https://github.com/Kwenta/token/blob/main/test/foundry/unit/TokenDistributor/TokenDistributor.t.sol) unit tests. I can update the Foundry unit tests to mirror changes to the new logic if you approve.

jcmonte commented on contracts/TokenDistributor.sol:
```diff
@@ -4,15 +4,17 @@ pragma solidity 0.8.19;
+    using BitMaps for BitMaps.BitMap;
+
+    /// @dev BitMap for storing claimed epochs
+    mapping(address to => BitMaps.BitMap claimedEpochs) private _claimedEpochsBitMap;  
```
I think @tommyrharper made these all public. Seems silly that anyone would want to view the bitmap, but private doesn't really afford any advantages here.

scorpion9979 replied:
I think the only advantage would be developer experience, knowing that this declaration is only needed for internal implementation and is not meant to be user-facing. I will still arrange to make this a public storage variable though.

scorpion9979 replied:
@jcmonte There turned out to be another reason why this needs to be either `private` or `internal`. Solidity does not allow internal or recursive types for public state variables.


![image](https://github.com/Kwenta/token/assets/4002635/63000f51-a3ea-4cfe-908e-5a25b7374332)

JaredBorders replied:
https://github.com/scorpion9979/kwenta-token/blob/816214d2f9d35503309b411877728e73726c90cd/contracts/TokenDistributor.sol#L183 should suffice

jcmonte commented on contracts/TokenDistributor.sol:
```diff
@@ -118,42 +120,48 @@ contract TokenDistributor is ITokenDistributor {
         _isEpochReady(epochNumber);
-        mapping(uint256 => bool) storage claimedEpochsTo = claimedEpochs[to];
-        if (claimedEpochsTo[epochNumber]) {
+        if (_claimedEpochsBitMap[to].get(epochNumber)) {
```
Can you not use your `claimedEpochs()` view function here?

scorpion9979 replied:
I think you meant to ask me to use `claimedEpochs()` function here. It's slightly less gas efficient to use it, but it's more dev friendly. I will make the change.

scorpion9979 replied:
Done.

jcmonte replied:
Yeah I think you can also follow the `_claimedEpochs()` internal `claimedEpochs()` external format if you really want the gas savings. But this is probably good enough

jcmonte commented on contracts/TokenDistributor.sol:
```diff
@@ -173,6 +181,11 @@ contract TokenDistributor is ITokenDistributor {
     }
 
+    /// @inheritdoc ITokenDistributor
+    function claimedEpochs(address to, uint epochNumber) public view override returns (bool) {
```
nit: `claimedEpoch` without the s

scorpion9979 replied:
Done.

jcmonte commented on contracts/TokenDistributor.sol:
```diff
@@ -118,42 +120,48 @@ contract TokenDistributor is ITokenDistributor {
             uint epochNumber = epochs[i];
-            _claimEpoch(to, epochNumber);
+            // TODO: possible optimization by placing the line below in unchecked block
+            // if we can guarantee `totalProportionalFees` will never overflow
```
unlikely to overflow because the rewardsToken will be USDC, but still a possibility for anyone using this contract with an arbitrary ERC20

would just leave as is, probably not worth gas savings

scorpion9979 replied:
I've removed the TODO.

jcmonte reviewed (changes requested):
Changes mostly look good, thanks!

jcmonte commented:
[PASS] testClaimMany() (gas: 447099)
[PASS] testClaimMultipleClaims() (gas: 657686)

Also could you change the name of these tests to include the # of claims made? Curious what the cost per claim is.

scorpion9979 commented:
> testClaimMultipleClaims

@jcmonte Will do! I would like to suggest to go for a more standardized way of naming Foundry tests in the repo. I'm personally a fan of [Matt Solomon's naming convention](https://twitter.com/PaulRBerg/status/1616068157088714765 ).

JaredBorders commented on test/foundry/unit/TokenDistributor/TokenDistributor.t.sol:
```diff
@@ -562,7 +562,7 @@ contract TokenDistributorTest is TokenDistributorSetup {
 
     /// @notice test claimMany
-    function testClaimMany() public {
+    function testClaimManyOnce() public {
```
Please add token balance assertions pre- and post-claim for user1 and the contract

JaredBorders commented on contracts/TokenDistributor.sol:
```diff
@@ -173,6 +179,11 @@ contract TokenDistributor is ITokenDistributor {
     }
 
+    /// @inheritdoc ITokenDistributor
+    function claimedEpoch(address to, uint epochNumber) public view override returns (bool) {
```
Please add a test for this function to assure it works as expected

JaredBorders reviewed (changes requested):
- incoming changes will likely reduce code coverage; new function that was introduced needs a test
- token balance assertions needed in test I commented on (should have already been present tbh 🫤)
- please update gas-snapshot in the repo (see https://github.com/scorpion9979/kwenta-token/blob/816214d2f9d35503309b411877728e73726c90cd/package.json#L30)

JaredBorders commented:
In the future, updating the `~/.gas-snapshot` instead of pasting changes in the PR description is preferable, but I'm still glad to see those optimizations!!

scorpion9979 commented:
@JaredBorders I've implemented the changes requested above.
Due to the following local error I encountered: `setup failed: No fork active`, I had to remove the following fork tests before running `npm run gas-snapshot`:
- test/foundry/integration/escrow.migrator.fork.t.sol
- test/foundry/integration/stakingV2.migration.fork.t.sol

Note that I had already setup the local `.env` file before running the command above.

JaredBorders commented:
> @JaredBorders I've implemented the changes requested above. Due to the following local error I encountered: `setup failed: No fork active`, I had to remove the following fork tests before running `npm run gas-snapshot`:
> 
> * test/foundry/integration/escrow.migrator.fork.t.sol
> * test/foundry/integration/stakingV2.migration.fork.t.sol
> 
> Note that I had already setup the local `.env` file before running the command above.

You will need to specify a fork env then. See here for reference: https://github.com/Kwenta/smart-margin-v3/blob/cacb85bd4c913a85c6bc978b1e8cde2585cee19a/package.json#L14C22-L14C90

scorpion9979 commented:
> > @JaredBorders I've implemented the changes requested above. Due to the following local error I encountered: `setup failed: No fork active`, I had to remove the following fork tests before running `npm run gas-snapshot`:
> > 
> > * test/foundry/integration/escrow.migrator.fork.t.sol
> > * test/foundry/integration/stakingV2.migration.fork.t.sol
> > 
> > Note that I had already setup the local `.env` file before running the command above.
> 
> You will need to specify a fork env then. See here for reference: https://github.com/Kwenta/smart-margin-v3/blob/cacb85bd4c913a85c6bc978b1e8cde2585cee19a/package.json#L14C22-L14C90

@JaredBorders I was able to spin the fork in the code itself and running all fork tests successfully by adding the following line in both files before `vm.rollFork`. I've removed my API key for obvious reasons. I think it could be beneficial to add the lines in the code, and also add the proper `.env` variables and config for Optimism mainnet.
I pushed another commit for updating the snapshot. The file should be all up-to-date now.

![image](https://github.com/Kwenta/token/assets/4002635/afca9bf8-c085-4d5a-9f39-be082ca3c199)
